ILogger_best_practice/kb/scenario_description/defaults.md

```diff
@@ -0,0 +1,182 @@
+
+The source generator is triggered when LoggerMessageAttribute is used on `partial` logging methods. When triggered, it is either able to autogenerate the implementation of the partial methods it's decorating, or produce compile-time diagnostics with hints about proper usage. The compile-time logging solution is typically considerably faster at run time than existing logging approaches. It achieves this by eliminating boxing, temporary allocations, and copies to the maximum extent possible.
+
+## Basic usage
+
+To use the `LoggerMessageAttribute`, the consuming class and method need to be `partial`. The code generator is triggered at compile time, and generates an implementation of the `partial` method.
+
+```csharp
+public static partial class Log
+{
+    [LoggerMessage(
+        Level = LogLevel.Information,
+        Message = "Could not open socket to `{HostName}`")]
+    public static partial void CouldNotOpenSocket(
+        ILogger logger, string hostName);
+}
+```
+
+In the preceding example, the logging method is `static` and the log level is specified in the attribute definition. When using the attribute in a static context, either the `ILogger` instance is required as a parameter, or modify the definition to use the `this` keyword to define the method as an extension method.
+
+```csharp
+public static partial class Log
+{
+    [LoggerMessage(
+        Level = LogLevel.Information,
+        Message = "Could not open socket to `{HostName}`")]
+    public static partial void CouldNotOpenSocket(
+        this ILogger logger, string hostName);
+}
+```
+
+You may choose to use the attribute in a non-static context as well. Consider the following example where the logging method is declared as an instance method. In this context, the logging method gets the logger by accessing an `ILogger` field in the containing class.
+
+```csharp
+public partial class InstanceLoggingExample
+{
+    private readonly ILogger _logger;
+
+    public InstanceLoggingExample(ILogger logger)
+    {
+        _logger = logger;
+    }
+
+    [LoggerMessage(
+        Level = LogLevel.Information,
+        Message = "Could not open socket to `{HostName}`")]
+    public partial void CouldNotOpenSocket(string hostName);
+}
+```
+
+Sometimes, the log level needs to be dynamic rather than statically built into the code. You can do this by omitting the log level from the attribute and instead requiring it as a parameter to the logging method.
+
+```csharp
+public static partial class Log
+{
+    [LoggerMessage(
+        Message = "Could not open socket to `{HostName}`")]
+    public static partial void CouldNotOpenSocket(
+        ILogger logger,
+        LogLevel level, /* Dynamic log level as parameter, rather than defined in attribute. */
+        string hostName);
+}
+```
+
+You can omit the logging message and `System.String.Empty` will be provided for the message. The state will contain the arguments, formatted as key-value pairs.
```

Comment:
If I'm not mistaken I think this triggers a build warning as well. We discussed getting rid of the warning but no consensus was reached so I believe it is still there.